parser/src/ast/expression.rs

```diff
@@ -621,6 +674,156 @@ impl SymbolAccess {
             ty: None,
         }
     }
+
+    /// Generates a new [SymbolAccess] that represents accessing this access, i.e.
+    /// nesting accesses. For example, if called with `AccessType::Index`, and
+    /// the current access type is `Default`, a new [SymbolAccess] is returned which
+    /// has an access type of `Index`. However, if the current access type was `Index`,
+    /// then the resulting [SymbolAccess] will have an access type of `Matrix`
+    ///
+    /// It is expected that the type of this access has been resolved already, and this
+    /// function will panic if that is not the case.
+    pub fn access(&self, access_type: AccessType) -> Result<Self, InvalidAccessError> {
+        let ty = self.ty.unwrap();
+        match self.access_type.clone() {
+            AccessType::Default => match access_type {
+                AccessType::Default => Ok(self.clone()),
+                AccessType::Index(idx) => match ty {
+                    Type::Felt => Err(InvalidAccessError::IndexIntoScalar),
+                    Type::Vector(len) if idx >= len => Err(InvalidAccessError::IndexOutOfBounds),
+                    Type::Vector(_) => Ok(Self {
+                        access_type: AccessType::Index(idx),
+                        ty: Some(Type::Felt),
+                        ..self.clone()
+                    }),
+                    Type::Matrix(rows, _) if idx >= rows => {
+                        Err(InvalidAccessError::IndexOutOfBounds)
+                    }
+                    Type::Matrix(_, cols) => Ok(Self {
+                        access_type: AccessType::Index(idx),
+                        ty: Some(Type::Vector(cols)),
+                        ..self.clone()
```

Comment:
nit: This function is really long and repeats very similar logic a few times. Is there any way to improve readability? Also, I think there's some logic that's repeated from the other `access` function. Can anything be consolidated?

Reply:
I'm not sure there is a meaningful way to break this up - we can't really escape the quadratic complexity of the function. Each combination of base access type, nested access type (relative to the type of the value produced by the base access) produces a unique result. 

Functions like this are annoying, since we must enumerate all of the combinations explicitly, but there also isn't really a better way to do it, at least in Rust with two enums involved. We could forcibly break it up by yanking parts out into new free functions, but I'm not sure that's actually easier to read. Better a long function that handles all of the cases clearly (and exposes any inherent patterns), than a short function that hides those details behind more abstraction - but that's a double-edged sword to be sure. Any thoughts on how you'd tackle it?

Reply:
Hm - my preference does tend to be using well-named helper functions and keeping the main function smaller, but I'd like to defer to @bobbinth on this again as well, since it's really a question of team style (similar to the `unwrap` discussion)

Reply:
I would agree with @grjte - would be better to have smaller, well-named helper functions which handle individual cases separately. There could be situations when this is not desirable - e.g., when there is a lot of context that needs to be passed into the helper functions. But if context is relatively small, I find that it usually improves readability (and reduces the amount of nesting).

For example, in this specific case, one way to do this could be to have a separate helper function for each initial match branch. Maybe something like this:
```Rust
match self.access_type.clone() {
    AccessType::Default => access_default(ty),
    AccessType::Slice(range) => access_slice(ty, range),
    AccessType::Index(base_idx) => access_index(ty, base_idx),
    AccessType::Matrix(_, _) => match access_type {
        AccessType::Default => Ok(self.clone()),
        _ => Err(InvalidAccessError::IndexIntoScalar),
    }
}
```
And then individual functions `access_default()`, `access_slice()`, `access_index()` (names could be different too) would be much easier to digest.